dendrite: Cleanup options, add global database option

tim-tx opened the pull request:
###### Description of changes

~~The `sync_api.search.enable` option is actually called `enabled` in dendrite's yaml config. https://github.com/matrix-org/dendrite/blob/39581af3ba657032fbf66ba3719a1cb334c0519b/dendrite-sample.monolith.yaml#L287~~

~~Using `enable` has no effect. Alternatively all of `sync_api.search` could be removed since I think the nixos module config has mostly only included options (w/ defaults) that are required for dendrite. Any non-required options are still supported since the module uses `pkgs.formats.yaml` generator.~~

EDIT:

  + Remove outdated `connection_string` options
  + Add `global.database.connection_string` option, force user to pick between global or sub-DBs, check that global DB is not SQLite
  + Remove non-essential options from yaml (also replace open registration flag with `extraFlags`)
  + Add defaults for `server_name` and `private_key` (required options for dendrite)
  + Generate private key if absent

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Mic92 commented on nixos/modules/services/matrix/dendrite.nix:
```diff
@@ -196,7 +196,7 @@ in
+          enabled = lib.mkEnableOption (lib.mdDoc "Dendrite's full-text search engine");
           index_path = lib.mkOption {
             type = lib.types.str;
             default = "${workingDir}/searchindex";
```
Do you see value in having this value for `index_path` being set? Or has it a good value already set by default?

winterqt replied:
Looking at it again, the default within Dendrite works fine for us because we set the working directory appropriately, so the default without this option definition is the same that it currently is. (Default in dendrite is `./searchindex`.)

I'm fine with dropping the search section entirely if that would be best, since all the defaults work for us.

Mic92 commented:
cc @winterqt who introduced this option.

winterqt commented:
Honestly, it looks like we could drop at least a few options that are currently present in the settings set, since Dendrite's defaults are the defaults we have defined, making ours redundant.

Is this something we should do, or no?

Mic92 commented:
> Honestly, it looks like we could drop at least a few options that are currently present in the settings set, since Dendrite's defaults are the defaults we have defined, making ours redundant.

Sounds good to me.

tim-tx committed:
dendrite: Clean up database options, add global database option

tim-tx committed:
dendrite: Add defaults for private_key and server_name, autogenerate private_key

tim-tx committed:
dendrite: Remove non-essential settings

tim-tx commented:
Dendrite needs at minimum the following config to start up:
```
global:
  server_name: something
  private_key: /path/to/private_key.pem
media_api
  base_path: /path/to/media_store
```
PLUS the `connection_string` keys, either the global one or all of the others for the different APIs. Keeping only required options in the module config, we could remove from `services.dendrite.settings`: 
  + `global.trusted_third_party_id_servers`
  + `sync_api.search`
  + `client_api.registration_disabled`

We could even remove all options in `settings` I guess, though maybe it's good to keep the required ones. Note that `client_api.registration_disabled` is true by default in dendrite, even if the key is omitted from the config.

tim-tx commented:
I updated this PR to include remove extra options as mentioned above. Maybe there is an argument for keeping the registration options since this is a very basic functionality, but I removed them anyway.

Also added a `settings.global.database.connection_string` option with logic to force the user to pick between the global one or all the other sub-DBs. Some of the sub-DBs were no longer used by dendrite so options for those were also removed.

Can certainly restructure this PR however is advised.

tim-tx commented:
Actually, I've had some trouble with `permission denied` errors when using the `loadCredential` option after switching to the `preStart` mechanism. Maybe related: https://github.com/systemd/systemd/issues/24291
